Share permissions per user, not per device

dborkan opened the pull request:
Share permissions per user, not per device.  Fixes https://github.com/uProxy/uproxy/issues/1014

Design doc: https://docs.google.com/document/d/1p16jeodYUYF9tp2mh9nDVIHlKagzHZLYVHRKZLZkspw/edit

* Consent.State is now part of the User object, not RemoteInstance.  This means that if I grant or request access I am granting/requesting for all instances of that user.
* remoteGrantsAccessToLocal has been removed from Consent.State.  Even though consent is now per user, when I (the local user) grant access to a remote user, I am granting access to all of that user's devices but only to my current device.  We now use the wireConsentFromRemote.isOffering on each instance to see which (if any) instances the user has permissions me from.
* All instance handshake messages now include consent (isOffering and isRequesting).  This allows us to remove an extra round trip of communication for each pair of instances.
* RemoteInstance.updateConsent has been merged into RemoteInstance.update, since we now always have consent fields with the instance handshake.  RemoteInstance.update now delays updating until onceLoaded is complete, and returns a Promise so that the caller can tell when update finishes (e.g. to notify the UI with the updated user and instances)
* The complex hidden expressions to toggle request/revoke/offer/etc buttons have been replaced by a simpler check for reading user.gettingConsentState and user.sharingConsentState enum values.  These are set in the UI.User's update method.

Tested with updated grunt test, running on Chrome and Firefox

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/1110)
<!-- Reviewable:end -->

dborkan commented:
@iislucas , @salomegeo , @lucyhe - you all may be interested in looking at these changes.  Thanks

iislucas commented:
I will review later today/tomorrow morning; thanks for getting this together so quickly!

jpevarnek commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1110)

Reviewed files:
- src/generic_core/consent.spec.ts @ r1
- src/generic_core/consent.ts @ r1
- src/generic_core/core.spec.ts @ r1
- src/generic_core/core.ts @ r1
- src/generic_core/local-instance.spec.ts @ r1
- src/generic_core/local-instance.ts @ r1
- src/generic_core/remote-instance.spec.ts @ r1
- src/generic_core/remote-instance.ts @ r1
- src/generic_core/social.spec.ts @ r1
- src/generic_core/social.ts @ r1
- src/generic_core/user.spec.ts @ r1
- src/generic_core/user.ts @ r1
- src/interfaces/instance.d.ts @ r1
- src/interfaces/network.d.ts @ r1
- src/interfaces/ui.d.ts @ r1
- src/uproxy.ts @ r1

---

As a heads up, I only scrolled through the code in `generic_ui`, so that definitely needs more eyes.

---

<sup>**[src/generic_core/consent.spec.ts, line 19 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYPbN2Xt56ieinibpO)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/consent.spec.ts#L19)):</sup>
Not actually relevant to this review: but this test looks a bit pointless in general.  We should have type-checking for this.

---

<sup>**[src/generic_core/consent.ts, line 11 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYnm4c_p1mhPswsV-0)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/consent.ts#L11)):</sup>
These comments feel overly verbose considering the variable names, could this possibly be done with a single comment at the top?

---

<sup>**[src/generic_core/core.spec.ts, line 19 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZHMrm5leDPQXiKs-f)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/core.spec.ts#L19)):</sup>
This is a bit outside the scope of this review (again), however, I think it might be better to either create a mock or use `createSpyObj` for this.  Feel free to ignore this.

---

<sup>**[src/generic_core/core.spec.ts, line 29 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYoSTEEUeeTR61Tdtj)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/core.spec.ts#L29)):</sup>
I think we can actually get rid of this entire block

---

<sup>**[src/generic_core/core.ts, line 422 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYrISQ3ZWXJ3HwQ_yE)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/core.ts#L422)):</sup>
We can replace the above with `var user = network.getUser(path.userId)`.  It also might be good to modify the definition of the `InstancePath` interface such that it extends `UserPath`

---

<sup>**[src/generic_core/remote-instance.spec.ts, line 23 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYtrF3AXwYJL1bbQ_U)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/remote-instance.spec.ts#L23)):</sup>
Personal preference: I think it looks better as `new Consent.State()`.  Feel free to ignore.

---

<sup>**[src/generic_core/remote-instance.spec.ts, line 79 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYuJ7DgCl4VIZMGtJH)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/remote-instance.spec.ts#L79)):</sup>
I don't think we have a strict guarantee that this will be run before the `then` block, do we?

---

<sup>**[src/generic_core/remote-instance.spec.ts, line 99 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYxbp4zCjNY2u08y8t)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/remote-instance.spec.ts#L99)):</sup>
Can you make `'instanceId1'` a constant? (it needs to be specifically used in some of the following tests)

---

<sup>**[src/generic_core/remote-instance.spec.ts, line 123 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYyM_mYdMZbOK507Xm)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/remote-instance.spec.ts#L123)):</sup>
Basing a test around whether we called a function to show a desktop notification to a user feels wrong.

If you want to make these tests based around that, I'd switch the descriptions to talk about showing a desktop notification instead of just notifying the UI (which is really just based on the variables changing as far as I understand Polymer)

---

<sup>**[src/generic_core/remote-instance.spec.ts, line 196 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYzabgv0m2m8Ztqzlx)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/remote-instance.spec.ts#L196)):</sup>
Not related to the review, but can you remove this log while here?

---

<sup>**[src/generic_core/remote-instance.ts, line 69 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ-qxrbsyauuJY8TzR)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/remote-instance.ts#L69)):</sup>
My impression of why this is being made public is so that you are able to call `stopShare` from `user.ts`.

My view of how this should fit together is that `RemoteInstance` should totally encapsulate `RemoteConnection` and should not need to expose it.  I think a much better solution would be to add a function to `RemoteInstance` that will close the underlying connection.

---

<sup>**[src/generic_core/social.spec.ts, line 107 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ3wtQ7JfFjWZnvcwX)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/social.spec.ts#L107)):</sup>
Why is a tick needed here?kkk

---

<sup>**[src/generic_core/user.spec.ts, line 176 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ4RsMgVS0NeCL8Z1c)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L176)):</sup>
I'd try to refer back to `instanceData` for these fields, or at least for `instanceId`

---

<sup>**[src/generic_core/user.spec.ts, line 193 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZIVWiSp8NptOlnLg7)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L193)):</sup>
General comment on this file: would it be possible to use `handleMessage` instead of `syncInstance_` throughout the tests?  `S

---

<sup>**[src/generic_core/user.spec.ts, line 227 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ6aAFJj79VS1e8PD3)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L227)):</sup>
Empty lines at the start of a functions looks a bit odd, I think it happens a few other places as well

---

<sup>**[src/generic_core/user.spec.ts, line 231 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ4ukC19Aoyj6-AxfJ)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L231)):</sup>
In these cases, I think it personally looks a bit cleaner (and makes indentation a bit less wild if you do:

```
user.modifyConsent(uProxy.ConsentUserAction.REQUEST).then(() => {
  expect(user.consent.localRequestsAccessFromRemote).toEqual(true);
  return user.modifyConsent(uProxy.ConsentUserAction.CANCEL_REQUEST);
}).then(() => {
  expect(user.consent.localRequestsAccessFromRemote).toEqual(false);
  done();
});
```

---

<sup>**[src/generic_core/user.spec.ts, line 240 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ5m1wvD7dgSFA5r5D)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L240)):</sup>
This does not seem to test anything the above test did not already handle.

---

<sup>**[src/generic_core/user.spec.ts, line 266 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ5TIo39xxwHDCbLA6)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L266)):</sup>
Same comment as above with the then blocks

---

<sup>**[src/generic_core/user.spec.ts, line 280 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ5ZVNH1HV7vhGMi2X)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L280)):</sup>
then block levels

---

<sup>**[src/generic_core/user.spec.ts, line 315 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ6yzvUT96gnu01s5A)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L315)):</sup>
I think a better title for this test would be something like "sets ignore request flag when remote is requesting" or something like that (I'd expect the test to function the same if you took out all the lines referring to `remoteRequestsAccessFromLocal` so feel like there should be some mention of that in the description)

---

<sup>**[src/generic_core/user.spec.ts, line 325 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ7YAnTVTH13Hj8o8I)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L325)):</sup>
And I just realized that all these tests share state... :(

I think we should definitely look at changing that, or at least not using the state for new tests being added.  Thoughts on that?

---

<sup>**[src/generic_core/user.spec.ts, line 358 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ7wotpVHDoTxB6XgO)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L358)):</sup>
This looks a bit funny indentation-wise.  At the very least, I'd recommend moving the `.then` call down a line (at best, same then indentation comment as above)

---

<sup>**[src/generic_core/user.ts, line 188 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ9YFBdorww77dDsOE)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L188)):</sup>
This should either check for an error or propogate the rejected promise (I'd personally be in favour of propogation, however, it does not seem like any of the errors from `syncInstance_` are critical enough to warrant propogation)

---

<sup>**[src/generic_core/user.ts, line 224 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ8ZhdjfT3OVJEt-Bq)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L224)):</sup>
Possible better name: `getLocalConsentBits`.  General feeling is this should encapsulate the fact that this only gets the flags describing the local state to the remote, not the other direction.

---

<sup>**[src/generic_core/user.ts, line 263 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ8wo3c91tX9MIvwNh)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L263)):</sup>
I think a better description here would be "Received an instance handshake from a client that is not online".  Thoughts?

Regardless, rejection messages should be wrapped in an `Error`.

---

<sup>**[src/generic_core/user.ts, line 289 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZAW8u1uoIhaviMPSz)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L289)):</sup>
Any thoughts on switching `saveToStorage` around to actually return a promise resolving once everything is successfully saved?kkk

---

<sup>**[src/generic_core/user.ts, line 318 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZBY9BnHUg-UeAP-lq)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L318)):</sup>
It feels like this function is doing a lot of processing that could be done in the UI based off of us just returning raw data.  Thoughts on moving towards that? (

---

<sup>**[src/generic_core/user.ts, line 357 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZBkjgG1FwW8qFghnD)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L357)):</sup>
Any idea if this is still relevant?

---

<sup>**[src/generic_core/user.ts, line 482 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZCKADW9wZX7p8uY9E)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L482)):</sup>
Not relevant to this review, however, this should at the very least return `Promise.resolve<void>()`.  This has the potential for a `TypeError` (kind of surprised that TypeScript does not catch this actually, we might want to look into that).

---

<sup>**[src/generic_core/user.ts, line 511 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZELCYmaCdLgGCATC0)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L511)):</sup>
I'm a bit unclear on what the promise returned by this command should reflect.  My personal impression of how it should work would be:

```
return this.onceLoaded.then(() => {
  /* error checking */
  return Promise.all([revoke access for everything])
}).then(() => {
  return Promise.all([send instance handshakes]);
}).then(() => {
  return this.saveToStorage();
}).then(() => {
  /* update the UI */
});
```

So that this returning would indicate that all steps had been successfully completed.

Anyways, obviously not saying it has to be like that, but it would be good to have a comment explaining what a successfully resolved promise indicated.

---

<sup>**[src/generic_core/user.ts, line 518 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZEfYX1fi3mn4Kjhp2)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L518)):</sup>
It seems like the function would return a rejected promise in this case instead of a successfully resolved one.

---

<sup>**[src/generic_core/user.ts, line 534 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZFNweVZHweeBi50f5)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L534)):</sup>
We are already in a block where `this.onceLoaded` has been called, it should not be needed here as well. (if it is, I think there is probably a bigger bug)

---

<sup>**[src/generic_ui/polymer/instance.ts, line 6 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZGDUzeGVC4HV8v7Ki)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_ui/polymer/instance.ts#L6)):</sup>
I'm guessing you meant to delete this.

---


<!-- Sent from Reviewable.io -->

lucyhe commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1110)

Reviewed files:
- src/generic_core/consent.spec.ts @ r1
- src/generic_core/consent.ts @ r1
- src/generic_core/core.spec.ts @ r1
- src/generic_core/core.ts @ r1
- src/generic_core/local-instance.spec.ts @ r1
- src/generic_core/local-instance.ts @ r1
- src/generic_core/remote-instance.spec.ts @ r1
- src/generic_core/remote-instance.ts @ r1
- src/generic_core/social.spec.ts @ r1
- src/generic_core/social.ts @ r1
- src/generic_core/user.spec.ts @ r1
- src/generic_core/user.ts @ r1
- src/generic_ui/polymer/contact.html @ r1
- src/generic_ui/polymer/contact.ts @ r1
- src/generic_ui/polymer/instance.html @ r1
- src/generic_ui/polymer/instance.ts @ r1
- src/generic_ui/polymer/roster.html @ r1
- src/generic_ui/polymer/roster.ts @ r1
- src/generic_ui/scripts/ui.spec.ts @ r1
- src/generic_ui/scripts/ui.ts @ r1
- src/generic_ui/scripts/user.spec.ts @ r1
- src/generic_ui/scripts/user.ts @ r1
- src/interfaces/instance.d.ts @ r1
- src/interfaces/network.d.ts @ r1
- src/interfaces/ui.d.ts @ r1
- src/uproxy.ts @ r1

---

Awesome! Getting the tests updated looks like it took a lot of work.. Aside from trying to make our tests independent (which Jonathan mentioned), looks solid! It's still kind of hard to keep track of what-means-what wrt consent, so I'll keep trying to think of clearer naming conventions..

---

<sup>**[src/generic_core/consent.ts, line 13 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jk_Ezzp0tU4HnyzfJt9)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/consent.ts#L13)):</sup>
Thoughts on naming this localInstanceGrantsAccessToRemote to differentiate it from localRequestsAccessFromRemote and remoteRequestsAccessFromLocal (implied to be user <--> user)? I think that also helps clarify why there's no remoteGrantsAccessToLocal in this class.

---

<sup>**[src/generic_core/core.ts, line 422 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYrISQ3ZWXJ3HwQ_yE)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/core.ts#L422)):</sup>
Did you mean this.getUser?

---

<sup>**[src/generic_core/remote-instance.spec.ts, line 69 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jk_SFHPsO5HgDfve5f3)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/remote-instance.spec.ts#L69)):</sup>
Does this depend on the onceLoaded from the previous test?

---

<sup>**[src/generic_core/remote-instance.spec.ts, line 112 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jk_d-KOySzi_E0o8Y64)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/remote-instance.spec.ts#L112)):</sup>
Could you add expect(instance.user. remoteRequestsAccessFromLocal).toEqual(false) at the beginning and .toEqual(true) in the then block?

---

<sup>**[src/generic_core/social.ts, line 131 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jk_hOzarn_Kg2d18vBx)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/social.ts#L131)):</sup>
might be nice to rename to localInstance

---

<sup>**[src/generic_core/user.ts, line 224 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ8ZhdjfT3OVJEt-Bq)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L224)):</sup>
Daniel + I were chatting about how this class can be renamed remote-user or something like that (preferably in another PR, since this one's getting pretty big). I think that would also make this more clear.

---

<sup>**[src/generic_core/user.ts, line 475 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jk_nkEL2hd3PMBlk-2E)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L475)):</sup>
Rename to localInstance?

---

<sup>**[src/generic_ui/polymer/contact.html, line 94 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jk_p44w06GbpARa4ibf)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_ui/polymer/contact.html#L94)):</sup>
This end tags are super helpful!

---

<sup>**[src/generic_ui/polymer/instance.html, line 68 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jk_pqaonakHC090XvqG)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_ui/polymer/instance.html#L68)):</sup>
typo

---

<sup>**[src/generic_ui/polymer/roster.html, line 70 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jk_qCKz6K4Ptv3H6fwp)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_ui/polymer/roster.html#L70)):</sup>
This fixes the 'same title in both tabs' bug, right?

---

<sup>**[src/generic_ui/scripts/user.spec.ts, line 27 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jk_wSl0Q1EyXsNfME-n)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_ui/scripts/user.spec.ts#L27)):</sup>
Did we want offeringInstances to be defined instead?

---


<!-- Sent from Reviewable.io -->

dborkan commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1110)







---

<sup>**[src/generic_core/consent.spec.ts, line 19 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYPbN2Xt56ieinibpO)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/consent.spec.ts#L19)):</sup>
I agree, this file is pretty much just a placeholder

---

<sup>**[src/generic_core/consent.ts, line 11 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYnm4c_p1mhPswsV-0)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/consent.ts#L11)):</sup>
I'm gonna keep them, they help me remember what this code is

---

<sup>**[src/generic_core/consent.ts, line 13 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jk_Ezzp0tU4HnyzfJt9)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/consent.ts#L13)):</sup>
Right now we aren't using "Instance" anywhere in the variable names.  Are you suggesting one start with "localInstance" and the other just start with "local"?  I'd rather keep these names, we have been using these names for many months and they are in the design documents which I don't can't edit

---

<sup>**[src/generic_core/core.spec.ts, line 29 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYoSTEEUeeTR61Tdtj)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/core.spec.ts#L29)):</sup>
This results in hiding any console.log calls, which doesn't hurt, plus its kinda irrelevant to the changes I made for consent

---

<sup>**[src/generic_core/remote-instance.spec.ts, line 69 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jk_SFHPsO5HgDfve5f3)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/remote-instance.spec.ts#L69)):</sup>
Yes, when I tried creating new instances I had really terrible async behavior that broke many of the following tests.  I'd rather keep it this way for now, we have several github issues already entered for cleaning up tests and can address dependencies between test cases then.

---

<sup>**[src/generic_core/remote-instance.spec.ts, line 79 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYuJ7DgCl4VIZMGtJH)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/remote-instance.spec.ts#L79)):</sup>
Yes we do - all then blocks get evaluated asynchronously, even if the promise is already resolved at the time you call .then.  This is nice because we can always our code will run async, rather than having to make sure it handles both sync and async cases.

---

<sup>**[src/generic_core/user.spec.ts, line 231 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ4ukC19Aoyj6-AxfJ)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L231)):</sup>
Sorry I prefer this way, I think it better shows that the 2nd call to user.modifyConsent is also async, and that the last expect statement is checking on the behavior of the 2nd modifyConsent call

---

<sup>**[src/generic_core/user.spec.ts, line 325 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ7YAnTVTH13Hj8o8I)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L325)):</sup>
I just copied all these tests from remote-instance.spec.ts, replace intsance.modifyConsent with user.modfiyConsent, and rewrote them to be async.  I agree they are bad but I'd rather leave them for now to keep this pull request a manageable size.  We can revisit these test cases later

---

<sup>**[src/generic_core/user.ts, line 224 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ8ZhdjfT3OVJEt-Bq)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L224)):</sup>
LocalConsentBits sounds wrong, this is actually the consent sent by the remote side to us.  Also "getConsentBits" is the name we've always been using, I just moved it from the remote-instance class to the user class

---

<sup>**[src/generic_core/user.ts, line 289 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZAW8u1uoIhaviMPSz)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L289)):</sup>
We intentionally don't wait for saveToStorage to finish so that we can notify the UI right away.  Even if it returned a promise, we'd just want to ignore that promise here

---

<sup>**[src/generic_core/user.ts, line 318 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZBY9BnHUg-UeAP-lq)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L318)):</sup>
Most of this logic has been here for months and is well tested.  The only new logic I added was send only instances which offer consent - this allows the UI do make simple checks in polymer using offeringInstances.length.  I added the isOnline field, which lets the UI know if there are online instances that don't offer access yet.  This code is all working well and I'd rather leave it as is than do another refactor now.

---

<sup>**[src/generic_core/user.ts, line 357 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZBkjgG1FwW8qFghnD)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L357)):</sup>
Probably not but I don't know enough about this to remove the comment.

---

<sup>**[src/generic_core/user.ts, line 475 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jk_nkEL2hd3PMBlk-2E)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L475)):</sup>
We've been using myInstance throughout the code since last summer, I'd rather keep this.  "local" only works well when constrasted with "remote", e.g. "localGrantsAccessToRemote" ("myGrantsAccessToRemote" doesn't work grammatically).  But otherwise I think "myInstance" is cleaner

---

<sup>**[src/generic_core/user.ts, line 511 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZELCYmaCdLgGCATC0)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L511)):</sup>
We don't actually need the Promise to resolve when all steps are completed, only when the user.consent fields are updated.  The only async bit we actually wait on here is this.onceLoaded - once that's complete we update permissions (using Consent.handleUserAction which is sync), stop sharing if needed (sync), send instance handshakes (sync), save (async but we don't care to wait on it), then notify the UI (sync)

---

<sup>**[src/generic_ui/polymer/roster.html, line 70 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jk_qCKz6K4Ptv3H6fwp)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_ui/polymer/roster.html#L70)):</sup>
Yes

---


<!-- Sent from Reviewable.io -->

jpevarnek commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1110)







---

<sup>**[src/generic_core/core.ts, line 422 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYrISQ3ZWXJ3HwQ_yE)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/core.ts#L422)):</sup>
Yes

---

<sup>**[src/generic_core/user.ts, line 224 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ8ZhdjfT3OVJEt-Bq)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L224)):</sup>
Okay, I'm now even more confused about what this function is returning, I think there should definitely be some clarification in the naming.  You are saying that `localRequestsAccessFromRemote`is whether, from the view of the remote user, the local user is requesting access?

---

<sup>**[src/generic_core/user.ts, line 511 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZELCYmaCdLgGCATC0)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L511)):</sup>
Okay, can you add a comment explaining what the semantics of returning from this is?

---


<!-- Sent from Reviewable.io -->

lucyhe commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1110)







---

<sup>**[src/generic_core/consent.ts, line 13 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jk_Ezzp0tU4HnyzfJt9)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/consent.ts#L13)):</sup>
Yp,! I meant 'localInstance' just the one for granting; since requesting is user-based but granting is still instance-based. 

---


<!-- Sent from Reviewable.io -->

salomegeo commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1110)







---

<sup>**[src/generic_core/core.spec.ts, line 29 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYoSTEEUeeTR61Tdtj)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/core.spec.ts#L29)):</sup>
No it's good to hide console.logs by default, it makes test output clearer. let's keep this

---

<sup>**[src/generic_core/remote-instance.ts, line 83 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkdhjnJ1iqJEsRmBNaO)** ([raw file](https://github.com/uproxy/uproxy/blob/89205078a7209982b3b2fbbffb2d4af991051511/src/generic_core/remote-instance.ts#L83)):</sup>
neat!

---

<sup>**[src/generic_core/remote-instance.ts, line 235 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkdvxLdU-1wlViKo2x8)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/remote-instance.ts#L235)):</sup>
Can you explain this? Isn't oldIsRequesting and newIsRequesting always the same?

---

<sup>**[src/generic_core/user.spec.ts, line 231 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ4ukC19Aoyj6-AxfJ)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L231)):</sup>
I agree with daniel on this, You'll see this style everywhere in our code base. We can discuss styling during our campfires.

---

<sup>**[src/generic_core/user.ts, line 255 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkdkLn_d4jO2C2QSGt-)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L255)):</sup>
Why does this need to return a promise?

---

<sup>**[src/generic_core/user.ts, line 333 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jkdu_ovPQcCY00HF4mz)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L333)):</sup>
You don't need totalInstanceCount and allInstanceIds, you can just use keys() and length, am I wrong?

---

<sup>**[src/generic_ui/polymer/roster.html, line 70 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jk_qCKz6K4Ptv3H6fwp)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_ui/polymer/roster.html#L70)):</sup>
sweet!

---


<!-- Sent from Reviewable.io -->

dborkan commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1110)







---

<sup>**[src/generic_core/consent.ts, line 13 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jk_Ezzp0tU4HnyzfJt9)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/consent.ts#L13)):</sup>
I'll update the comment to make this clear

---

<sup>**[src/generic_core/core.spec.ts, line 29 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYoSTEEUeeTR61Tdtj)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/core.spec.ts#L29)):</sup>
I just removed this along with the trace from MockStorage, now things are cleaner and we don't need to spy on console.log

---

<sup>**[src/generic_core/core.ts, line 422 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYrISQ3ZWXJ3HwQ_yE)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/core.ts#L422)):</sup>
Done

---

<sup>**[src/generic_core/remote-instance.spec.ts, line 23 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYtrF3AXwYJL1bbQ_U)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/remote-instance.spec.ts#L23)):</sup>
Done.

---

<sup>**[src/generic_core/remote-instance.spec.ts, line 99 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYxbp4zCjNY2u08y8t)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/remote-instance.spec.ts#L99)):</sup>
Done.

---

<sup>**[src/generic_core/remote-instance.spec.ts, line 112 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jk_d-KOySzi_E0o8Y64)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/remote-instance.spec.ts#L112)):</sup>
Done.

---

<sup>**[src/generic_core/remote-instance.spec.ts, line 123 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYyM_mYdMZbOK507Xm)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/remote-instance.spec.ts#L123)):</sup>
Just updated the description.  We haven't been generating notification text in the core for a long time, although in the future it might be nicer (and make for better tests) if we generate notification text in the UI instead based on some other update from the core.

---

<sup>**[src/generic_core/remote-instance.spec.ts, line 196 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkYzabgv0m2m8Ztqzlx)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/remote-instance.spec.ts#L196)):</sup>
Done.

---

<sup>**[src/generic_core/remote-instance.ts, line 69 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ-qxrbsyauuJY8TzR)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/remote-instance.ts#L69)):</sup>
Done.

---

<sup>**[src/generic_core/remote-instance.ts, line 235 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkdvxLdU-1wlViKo2x8)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/remote-instance.ts#L235)):</sup>
No - this.user.updateRemoteRequestsAccessFromLocal() can modify that value.  i.e. if you just got an instance handshake that set isRequesting=false and that had been the only instance requesting access, user.consent.remoteRequestsAccessFromLocal will change from true to false.  But if there are other instances still requesting access, user.consent.remoteRequestsAccessFromLocal will remain true

---

<sup>**[src/generic_core/social.spec.ts, line 107 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ3wtQ7JfFjWZnvcwX)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/social.spec.ts#L107)):</sup>
Removed, not needed after all

---

<sup>**[src/generic_core/social.ts, line 131 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jk_hOzarn_Kg2d18vBx)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/social.ts#L131)):</sup>
See comment below - we'll keep this as myInstance for now

---

<sup>**[src/generic_core/user.spec.ts, line 176 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ4RsMgVS0NeCL8Z1c)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L176)):</sup>
Done.

---

<sup>**[src/generic_core/user.spec.ts, line 193 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZIVWiSp8NptOlnLg7)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L193)):</sup>
Just discussed this with Jonathan, we are going to leave this as-is for now.  The problem is that syncInstance_ returns a Promise, but handleMessage does not (handleMessage doesn't really care about the result of syncInstance).  Making handleMessage return a Promise would be a major pain and only really help a bit in making the tests not call private methods

---

<sup>**[src/generic_core/user.spec.ts, line 227 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ6aAFJj79VS1e8PD3)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L227)):</sup>
Done.

---

<sup>**[src/generic_core/user.spec.ts, line 231 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ4ukC19Aoyj6-AxfJ)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L231)):</sup>
Discussed with Jonathan, we'll leave this as is for now.

---

<sup>**[src/generic_core/user.spec.ts, line 240 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ5m1wvD7dgSFA5r5D)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L240)):</sup>
Removed

---

<sup>**[src/generic_core/user.spec.ts, line 315 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ6yzvUT96gnu01s5A)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L315)):</sup>
Just changed the title to "ignoring request does not change remoteRequestsAccessFromLocal"

---

<sup>**[src/generic_core/user.ts, line 224 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ8ZhdjfT3OVJEt-Bq)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L224)):</sup>
Just removed getConsentBits - it was only called in one place (sendInstanceHandshake) and it's clearer to inline this object there

---

<sup>**[src/generic_core/user.ts, line 255 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkdkLn_d4jO2C2QSGt-)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L255)):</sup>
syncInstance_ ends by calling instance.update which is now async, as it doesn't complete until instance.onceLoaded resolves (we don't want to update consent fields only to have them clobbered when storage load completes).  No real callers check the Promise returned by syncInstance_ but this is needed for testers to check expectations after syncInstance_

---

<sup>**[src/generic_core/user.ts, line 263 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ8wo3c91tX9MIvwNh)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L263)):</sup>
We use this wording because the client could be either OFFLINE or ONLINE_WITH_OTHER_APP (e.g. if they are a GTalk user sending us an XMPP message from the chat client in GMail).

But I did change this to an error

---

<sup>**[src/generic_core/user.ts, line 333 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jkdu_ovPQcCY00HF4mz)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L333)):</sup>
Done.

---

<sup>**[src/generic_core/user.ts, line 511 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZELCYmaCdLgGCATC0)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L511)):</sup>
Just refactored a bit to make it more clear

---

<sup>**[src/generic_core/user.ts, line 518 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZEfYX1fi3mn4Kjhp2)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L518)):</sup>
Done.

---

<sup>**[src/generic_core/user.ts, line 534 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZFNweVZHweeBi50f5)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L534)):</sup>
Done.

---

<sup>**[src/generic_ui/polymer/instance.html, line 68 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jk_pqaonakHC090XvqG)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_ui/polymer/instance.html#L68)):</sup>
Done.

---

<sup>**[src/generic_ui/polymer/instance.ts, line 6 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZGDUzeGVC4HV8v7Ki)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_ui/polymer/instance.ts#L6)):</sup>
Done.

---

<sup>**[src/generic_ui/scripts/user.spec.ts, line 27 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jk_wSl0Q1EyXsNfME-n)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_ui/scripts/user.spec.ts#L27)):</sup>
Done.

---


<!-- Sent from Reviewable.io -->

dborkan commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1110)



---

I believe 

---

<sup>**[src/generic_core/user.spec.ts, line 280 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ5ZVNH1HV7vhGMi2X)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L280)):</sup>
discussed this, will leave as is

---

<sup>**[src/generic_core/user.ts, line 188 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ9YFBdorww77dDsOE)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.ts#L188)):</sup>
Done

---


<!-- Sent from Reviewable.io -->

dborkan commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1110)







---

<sup>**[src/generic_core/core.spec.ts, line 19 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZHMrm5leDPQXiKs-f)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/core.spec.ts#L19)):</sup>
Ack, will ignore for now

---


<!-- Sent from Reviewable.io -->

dborkan commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1110)







---

<sup>**[src/generic_core/user.spec.ts, line 266 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ5TIo39xxwHDCbLA6)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L266)):</sup>
Discussed this with Jonathan, will leave as is

---

<sup>**[src/generic_core/user.spec.ts, line 358 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-JkZ7wotpVHDoTxB6XgO)** ([raw file](https://github.com/uproxy/uproxy/blob/619a421abcb2b615274ae8e0ebeed545f55de52c/src/generic_core/user.spec.ts#L358)):</sup>
Will leave indentation as is for now

---


<!-- Sent from Reviewable.io -->

dborkan commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1110)



---

(sorry hit publish too soon before)

I believe I've addressed all the comments.  PTAL, we want to have this merged tomorrow morning if possible.  Thanks!!

---



<!-- Sent from Reviewable.io -->

salomegeo commented:
:+1:

lucyhe commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1110)

Reviewed files:
- src/generic_core/consent.ts @ r2
- src/generic_core/core.ts @ r2
- src/generic_core/remote-instance.spec.ts @ r2
- src/generic_core/remote-instance.ts @ r2
- src/generic_core/social.spec.ts @ r2
- src/generic_core/user.spec.ts @ r2
- src/generic_core/user.ts @ r2
- src/generic_ui/polymer/instance.html @ r2
- src/generic_ui/polymer/instance.ts @ r2
- src/generic_ui/scripts/user.spec.ts @ r2
- src/interfaces/instance.d.ts @ r2
- src/mocks/freedom-mocks.ts @ r2
- src/uproxy.ts @ r2

---

:+1:

---



<!-- Sent from Reviewable.io -->

jpevarnek commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1110)

Reviewed files:
- src/generic_core/consent.ts @ r2
- src/generic_core/core.ts @ r2
- src/generic_core/remote-instance.spec.ts @ r2
- src/generic_core/remote-instance.ts @ r2
- src/generic_core/social.spec.ts @ r2
- src/generic_core/user.spec.ts @ r2
- src/generic_core/user.ts @ r2
- src/interfaces/instance.d.ts @ r2
- src/mocks/freedom-mocks.ts @ r2
- src/uproxy.ts @ r2

---

One comment from me, then LGTM

---

<sup>**[src/generic_core/user.spec.ts, line 19 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1110#-Jkmb_nwgM4jabP4sFPa)** ([raw file](https://github.com/uproxy/uproxy/blob/3ecf94cf2d985042beef9318f971683c11d57f20/src/generic_core/user.spec.ts#L19)):</sup>
Can you delete these lines completely?

---


<!-- Sent from Reviewable.io -->
